### PR TITLE
【Hackathon 5th No.22】Add CosineAnnealingWarmRestarts API to Paddle -part

### PR DESCRIPTION
**Pull request by Patrick-Star125**
### PR types

New features

### PR changes

APIs

### Description

Add CosineAnnealingWarmRestarts API to Paddle

### Link

Rfc PR: https://github.com/PaddlePaddle/community/pull/649
docs PR: https://github.com/PaddlePaddle/docs/pull/6286

待完成

- [x] RFC文档修改
- [ ] 中文文档

### CODE REVIEW

**Comment by luotao1** [comment]
请解决下冲突并通过多条CI

File: test/legacy_test/test_lr_scheduler.py [hunk 1]
```diff
@@ -214,6 +214,203 @@ def _test_dygraph(self, place, kwargs):
...
+            )
+
+        places = [paddle.CPUPlace()]
+        if core.is_compiled_with_cuda():
+            places.append(paddle.CUDAPlace(0))
+
+        for place in places:
+            kwargs = {
+                'learning_rate': 0.5,
+                'T_0': 1,
```
Comment (jeff41404) [inline test/legacy_test/test_lr_scheduler.py]: `T_0 = 1` is special case, should add more test case when `T_0 > 1`

**Comment by jeff41404** [comment]
find a clerical error in [rfc](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20230925_api_design_for_CosineAnnealingWarmRestarts.md?plain=1#L146), `leaning_rate` should be `learning_rate`

File: test/legacy_test/test_lr_scheduler.py (hunk 1 above)
Comment (Patrick-Star125) [reply test/legacy_test/test_lr_scheduler.py]: Done

**Review by jeff41404 — DISMISSED** [review]
LGTM

File: python/paddle/optimizer/lr.py [hunk 2]
```diff
@@ -2349,6 +2350,169 @@ def get_lr(self):
...
+
+    .. math::
+        \eta_t = \eta_{min} + \frac{1}{2}(\eta_{max} - \eta_{min})\left(1 +
+        \cos\left(\frac{T_{cur}}{T_{i}}\pi\right)\right)
+
+    When :math:`T_{cur}=T_{i}`, set :math:`\eta_t = \eta_{min}`.
+    When :math:`T_{cur}=0` after restart, set :math:`\eta_t=\eta_{max}`.
+
+    It has been proposed in
+    `SGDR: Stochastic Gradient Descent with Warm Restarts`_.
```
Comment (sunzhongkai588) [inline python/paddle/optimizer/lr.py]: 链接没附上吧。写法参考[引用方式](https://github.com/PaddlePaddle/docs/wiki/飞桨文档相互引用#情况二a文档为rst格式b文档为非rst格式)

File: python/paddle/optimizer/lr.py [hunk 3]
```diff
@@ -2349,6 +2350,169 @@ def get_lr(self):
...
+
+    Args:
+        learning_rate (float): Initial learning rate.
+        T_0 (int): Number of iterations for the first restart.
+        T_mult (int, optional): A factor increases :math:`T_{i}` after a restart. Default: 1.
+        eta_min (float, optional): Minimum learning rate. Default: 0.
+        last_epoch (int, optional): The index of last epoch. Default: -1.
```
Comment (sunzhongkai588) [inline python/paddle/optimizer/lr.py]: 补充默认值代表的含义

**Comment by Patrick-Star125** [comment]
Done

**Review by sunzhongkai588 — APPROVED** [review]
LGTM for docs，请补充中文文档

**Pull request merged**